### DISCUSSION TITLE
How to Accelerate Computing with GPU

### DISCUSSION THREAD

**Discussion by fanyimeng0** [NVlabs/sionna#481] [General]
I've been using Sionna for some simulations lately, but it seems that the computational power of the GPU is not being used to its full potential. This seems to be due to the fact that I'm using some other libraries to do the calculations, e.g., from skimage.feature import peak_local_max, and when using these, I'm converting the TensorFlow variables to NumPy and using a for loop instead of batch processing.

I have learned a lot from the [Sionna discussion on GitHub](https://github.com/NVlabs/sionna/discussions/10#discussioncomment-2711186) and the [Discover Sionna guide](https://nvlabs.github.io/sionna/examples/Discover_Sionna.html#Sionna-Data-flow-and-Design-Paradigms). However, I still can't solve my problem.

Does this mean that I can't use the GPU efficiently unless I completely rewrite some of my NumPy-related functions to TensorFlow's functions? If you have some development experience in this area, I hope you can provide some ideas. Thanks a lot!

**Comment by SebastianCa** [comment]
Hi @fanyimeng0,
That is correct. You need to implement your experiments with a library that has GPU support. For Sionna, we have chosen TensorFlow as the backend (and Dr.Jit for the raytracing part). Therefore, I recommend using TensorFlow for seamless integration with Sionna.

**Reply by fanyimeng0** [reply]
Thanks for your response. It really helps.

**Reply by merlinND** [reply]
@fanyimeng0 For deeper performance analysis, you can have a look at [NSight Systems](https://developer.nvidia.com/nsight-systems) and NVTX to track down where the time is spent and whether the GPU is busy or not at any given time during execution.
But @SebastianCa's suggestion of moving CPU-side processing to the GPU is indeed likely to have the biggest impact on performance.
